python/lsst/daf/butler/registry/databases/sqlite.py

```diff
@@ -106,34 +107,133 @@ class SqliteDatabase(Database):
 
     Parameters
     ----------
+    cs : `ConnectionStruct`
+        An existing connection created by a previous call to `connect`.
     origin : `int`
         An integer ID that should be used as the default for any datasets,
         quanta, or other entities that use a (autoincrement, origin) compound
         primary key.
-    filename : `str`, optional
-        Absolute or relative path to the database file to load, or `None` to
-        create a temporary in-memory database.
+    namespace : `str`, optional
+        The namespace (schema) this database is associated with.  If `None`,
+        the default schema for the connection is used (which may be `None`).
     writeable : `bool`, optional
-        If `True` (default) allow writes to the database.
+            If `True`, allow write operations on the database, including
+            ``CREATE TABLE``.
+
+    Notes
+    -----
+    The case where ``namespace is not None`` is not yet tested, and may be
+    broken; we need an API for attaching to different databases in order to
+    write those tests, but haven't yet worked out what is common/different
+    across databases well enough to define it.
     """
 
-    def __init__(self, *, origin: int, filename: Optional[str] = None, writeable: bool = True):
-        target = f"file:{filename}" if filename is not None else ":memory:"
-        if not writeable:
-            target += '?mode=ro&uri=true'
+    def __init__(self, *, cs: ConnectionStruct, origin: int, namespace: Optional[str] = None,
+                 writeable: bool = True):
+        super().__init__(origin=origin, cs=cs, namespace=namespace)
+        # Get the filename from a call to 'PRAGMA database_list'.
+        with closing(cs.connection.connection.cursor()) as cursor:
+            dbList = list(cursor.execute("PRAGMA database_list").fetchall())
+        if len(dbList) == 0:
+            raise RuntimeError("No database in connection.")
+        for _, dbname, filename in dbList:
```

Comment:
I think you can simplify conditional logic below if you do `namespace = "main" if namespace is None else namespace`